src/ArduinoCloudProperty.hpp

```diff
@@ -98,68 +146,66 @@ class ArduinoCloudProperty {
     bool shouldBeUpdated();
     void execCallbackOnChange();
     void execCallbackOnSync();
-    void forceCallbackOnChange();
-    void setPreviousCloudChangeTimestamp(unsigned long cloudChangeTime);
     void setLastCloudChangeTimestamp(unsigned long cloudChangeTime);
     void setLastLocalChangeTimestamp(unsigned long localChangeTime);
-    unsigned long getPreviousCloudChangeTimestamp();
     unsigned long getLastCloudChangeTimestamp();
     unsigned long getLastLocalChangeTimestamp();
-    void setPropertyValue(T const val);
-    void setCloudShadowValue(T const val);
-    T    getCloudShadowValue();
-    void setLocalShadowValue(T const val);
-    T    getLocalShadowValue();
-    void updateTime(unsigned long changeEventTime);
-    bool isChangedLocally();
+
+    void updateLocalTimestamp();
     void append(CborEncoder * encoder);
+    void appendAttributeReal(bool value, String attributeName = "");
+    void appendAttributeReal(int value, String attributeName = "");
+    void appendAttributeReal(float value, String attributeName = "");
+    void appendAttributeReal(String value, String attributeName = "");
+    void appendAttributeName(String attributeName, std::function<void (CborEncoder& mapEncoder)>f);
+    void setAttributesFromCloud(LinkedList<CborMapData *> *map_data_list);
+    void setAttributeReal(bool& value, String attributeName = "");
+    void setAttributeReal(int& value, String attributeName = "");
+    void setAttributeReal(float& value, String attributeName = "");
+    void setAttributeReal(String& value, String attributeName = "");
+    void setAttributeReal(String attributeName, std::function<void (CborMapData *md)>setValue);
+    String getAttributeName(String propertyName, char separator);
+
+    virtual bool isDifferentFromCloud() = 0;
+    virtual void fromCloudToLocal() = 0;
+    virtual void fromLocalToCloud() = 0;
+    virtual void appendAttributesToCloud() = 0;
+    virtual void setAttributesFromCloud() = 0;
+    virtual bool isPrimitive() {
```

Comment:
I suggest to change `isPrimitive()` into a pure virtual function. If derived classes fail to implement this function an object of the derived class can not be instantiated. This way future programmers are forced to consider the implementation of this function when adding another type and it can not happen that a false default value is returned (which is a very probable error case otherwise).

Reply:
I defined this function in order to manage the 4 primitive wrappers. Any other property type will not be primitive, isPrimitive() must return false for any other implementation. No need to see/redefine this method in no primitive types.